dali/pipeline/operators/reader/loader/video_loader.cc

```diff
@@ -356,54 +403,67 @@ void VideoLoader::read_file() {
           continue;
       }
 
-      auto frame = av_rescale_q(pkt->pts,
+      auto frame = av_rescale_q(pkt->pts - file.start_time_,
                                 file.stream_base_,
                                 file.frame_base_);
       LOG_LINE << "Frame candidate " << frame << " (for " << req.frame  <<" )...\n";
 
       file.last_frame_ = frame;
-      auto key = pkt->flags & AV_PKT_FLAG_KEY;
+      key = (pkt->flags & AV_PKT_FLAG_KEY) != 0;
+      int pkt_frames = 1;
+      if (pkt->duration) {
+        pkt_frames = av_rescale_q(pkt->duration, file.stream_base_, file.frame_base_);
+        LOG_LINE << "Duration: " << pkt->duration
+                 << "\nPacket contains " << pkt_frames << " frames\n";
+      }
 
-      if (frame >= req.frame) {
+      if (frame > req.frame) {
         if (key) {
-          static auto final_try = false;
-          if (frame > req.frame + nonkey_frame_count) {
+          if (frames_left <= 0)
+            break;
+          if (is_first_frame) {
             LOG_LINE << device_id_ << ": We got ahead of ourselves! "
                           << frame << " > " << req.frame << " + "
                           << nonkey_frame_count
                           << " seek_hack = " << seek_hack << std::endl;
-            seek_hack *= 2;
-            if (final_try) {
+            if (seek_must_succeed) {
               std::stringstream ss;
               ss << device_id_ << ": failed to seek frame "
                   << req.frame;
               DALI_FAIL(ss.str());
             }
             if (req.frame > seek_hack) {
               seek(file, req.frame - seek_hack);
+              seek_hack *= 2;
             } else {
-              final_try = true;
+              seek_must_succeed = true;
               seek(file, 0);
             }
             continue;
           } else {
-            req.frame += nonkey_frame_count + 1;
-            req.count -= nonkey_frame_count + 1;
             nonkey_frame_count = 0;
           }
-          final_try = false;
+          seek_must_succeed = false;
         } else {
-          nonkey_frame_count++;
+          nonkey_frame_count += pkt_frames;
           // A heuristic so we don't go way over... what should "20" be?
-          if (frame > req.frame + req.count + 20) {
-            // This should end the loop
-            req.frame += nonkey_frame_count;
-            req.count -= nonkey_frame_count;
-            nonkey_frame_count = 0;
+          if (frames_left <= 0 && frame > req.frame + req.count + 20) {
```

Comment:
if we move the `frames_left > 0` check to the `while` condition you could remove it from here